client/shared/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -289,16 +291,57 @@ object Client {
       go(source).stream
     }
 
-    def runOrDispose(req: Request[F]): F[Resource[F, Response[F]]] =
-      Ref[F].of(false).map { disposed =>
-        val req0 = addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
-        Resource
-          .eval(app(req0))
-          .onFinalize(disposed.set(true))
-          .map(_.pipeBodyThrough(until(disposed)))
-      }
+    def run(
+        message: Message[F],
+        refOp: Option[Ref[F, Boolean]] = None,
+    ): Resource[F, Response[F]] = message match {
+      case req: Request[F] =>
+        Resource.eval(Ref[F].of(false)).flatMap { disposed =>
+          val reqAugmented =
+            addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
+          Resource
+            .eval(app(reqAugmented))
+            .onFinalize(disposed.set(true))
+            .flatMap(run(_, Option(disposed)))
+        }
+      case resp: Response[F] =>
+        for {
+          disposed <- Resource.eval(refOp.fold(Ref[F].of(false))(F.pure))
+          channel <- Resource.eval(Channel.synchronous[F, Chunk[Byte]])
```

Comment:
Everything looks great and the tests you added are excellent. Along those lines, if you don't mind, I have one last suggestion: can we add a test that will fail if this is switched to a `Channel.unbounded`? Something to verify that the body is not loaded into memory.

Reply:
Nice idea! The test intercepts the moment the publication is completed before the response body is read by the client. Unfortunate, I had to use `IO.sleep` there to wait when the publication will be completed. I couldn't come up with any better approach to handle this. The logic might wait for a value of the `completed` to be true, but it can cause excessive waiting in the case when the logic uses `Channel.synchronous`.

Reply:
> The logic might wait for a value of the `completed` to be true, but it can cause excessive waiting in the case when the logic uses `Channel.synchronous`.

Oh, actually that might be another good approach. I added some suggestions about using the `executeEmbed` test runtime which uses simulated time. With that one, you can actually verify that an `IO` never completes. So maybe we can write the test like that?

Reply:
Thanks! Now, it's better.